internal/pubsub/samples/pubsub_test.go

```diff
@@ -0,0 +1,91 @@
+package pubsub_test
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"os/exec"
+	"sort"
+	"strings"
+	"sync"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+var topicFlag = flag.String("topic", "test-topic", "topic to be used for testing pub and sub commands")
+var subFlag = flag.String("sub", "test-subscription-1", "subscription to be used for testing pub and sub commands")
+var envsFlag = flag.String("pubsub_envs", "gcp,rabbit", "environments in which to run the test")
+
+func TestMain(m *testing.M) {
+	flag.Parse()
+	os.Exit(m.Run())
+}
+
+func TestPubAndSubCommands(t *testing.T) {
+	if !commandExists("pub") {
```

Comment:
Couldn't you just let runWithInput or run fail?

Reply:
That would break the build on Travis and for anyone who hasn't compiled the `pub` and `sub` example programs.

Reply:
This tests the logic in gcmsg/main.go. I feel I may not be quite understanding the question.

Reply:
I changed the test to no longer rely on the `gcmsg` command being on the `$PATH` and now your point seems clearer. We could get most of the value from the test by extracting the URL parsing logic and writing unit tests for that. Then we wouldn't need any record/replay on it.

Reply:
I was making a higher-level point: tests have maintenance cost, too. Ideally, this sample should be straightforward enough that basic manual usage and code review should be enough to have confidence that the sample works. If there is logic in the sample complex enough that tests are buying us confidence in its correctness, then that's signal that the API may be deficient. If the tests never run on Travis, then we don't receive much benefit from them and we incur maintenance cost.

(The only other places in Go Cloud that I'm aware of right now that have tests that don't run on Travis are the server and SQL dialer packages, and this is because record/replay is infeasible due to their nature. It would bring significantly more benefit if they ran regularly, but that brings other costs.)

Reply:
I think I generally agree with your point here.

The URL parsing logic is complex enough to need tests. This logic is only provisionally meant to live in the sample. The plan is to move it into the pubsub API if users say they want that, or if we decide it makes sense for some other reason such as consistency with blob's URL functionality.